### ISSUE TITLE
-force cuDNN fails 

### ISSUE THREAD

**Issue by slayerduck** [nagadomi/waifu2x#128]
Hi this is the hoster of waifu2x.booru.pics. Just now i updated to the latest version of waifu2x. It works, however I'd like to use cuDNN to speed it up. 

Trying to use -force_cudnn 1 fails, error: 

th waifu2x.lua -m noise -noise_level 1 -i 1.jpg -o output_image.png -force_cudnn 1
/home/waifu/torch/bin/luajit: lib/w2nn.lua:26: attempt to index global 'cudnn' (a nil value)
stack traceback:
        lib/w2nn.lua:26: in function 'load_model'
        waifu2x.lua:62: in function 'convert_image'
        waifu2x.lua:283: in function 'waifu2x'
        waifu2x.lua:288: in main chunk
        [C]: in function 'dofile'
        /home/waifu/torch/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
        [C]: at 0x00406010

I have cuDNN lib (libcudnn.so) installed at /usr/local/cuda/lib64 and torch cuDNN is installed to (not sure if its still needed?)

Paths inside my bashrc:
export CUDA_BIN_PATH=/usr/local/cuda
export LD_LIBRARY_PATH=:/usr/local/cuda/lib64
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib

Before the new version cuDNN was always needed but now that its optional installation documents are lacking. Please advice how i can get it running with cuDNN :)

**Comment by nagadomi** [comment]
cudnn.torch(lua package) is required.
```
luarocks install cudnn
```

**Comment by slayerduck** [comment]
torch cuDNN was already installed for previous versions of waifu2x, i ran "luarocks install cudnn" again and it installed without error. waifu2x.lua produces the same error as mentioned in OP.

**Comment by nagadomi** [comment]
Maybe CUDA 7.5 and cuDNN 5.0(or 5.1) is required. Which version did you use?

What happens when you run the following command?
```
th -e "require 'cudnn'"
```

**Comment by slayerduck** [comment]
I downloaded cudnn-7.5-linux-x64-v5.1-rc.tgz 

th -e "require 'cudnn'" has no output, so i assume it worked?

**Comment by nagadomi** [comment]
I think so.

**Comment by nagadomi** [comment]
Are you still getting error? In your env, `require 'cudnn'` is successful but `cudnn` is not defined. I am not sure what happened. 

I noticed that your CUDA_BIN_PATH is invalid. (but I think it is not related)
Here is my env variables and update command that I used.
```
# I installed cudnn to /usr/local/cuda/lib64/libcudnn.so, /usr/local/cuda/include/cudnn.h
export PATH=/usr/local/cuda/bin:$PATH
export C_INCLUDE_PATH=/usr/local/cuda/include:$C_INCLUDE_PATH
export CPLUS_INCLUDE_PATH=/usr/local/cuda/include:$CPLUS_INCLUDE_PATH
export LD_LIBRARY_PATH=/usr/local/cuda/lib64:$LD_LIBRARY_PATH
export CUDA_BIN_PATH=/usr/local/cuda/bin
```
```
source ~/.zshrc # if you using bash, this line is ` source ~/.bashrc`.
```
```
#!/bin/sh
# update torch core modules
luarocks install cwrap
luarocks install paths
luarocks install torch
luarocks install trepl
luarocks install sys
luarocks install xlua
luarocks install nn
luarocks install optim
luarocks install image
luarocks install graphicsmagick
luarocks install cutorch
luarocks install cunn
luarocks install cudnn
```

**Comment by slayerduck** [comment]
Updated env variables with the ones you given. Confirmed /usr/local/cuda/lib64/libcudnn.so and /usr/local/cuda/include/cudnn.h exist. Ran updates on all torch modules. 

Still getting the same error :(

**Comment by nagadomi** [comment]
What happens when you run the following command?
```
$ th
th> require 'cudnn'
```
Here is the results on my server.
```
{
  SpatialCrossEntropyCriterion : {...}
  VolumetricConvolution : {...}
  SpatialAveragePooling : {...}
  C : userdata: 0x4056a528
  benchmark : false
  ClippedReLU : {...}
  _Pooling : {...}
  SpatialBatchNormalization : {...}
  fastest : false
  toDescriptor : function: 0x4056ac18
  ReLU : {...}
  SpatialCrossMapLRN : {...}
  convert : function: 0x40574830
  version : 5005
  SoftMax : {...}
  TemporalConvolution : {...}
  getHandle : function: 0x4056a540
  SpatialLogSoftMax : {...}
  getSharedWorkspace : function: 0x4056acd0
  _Pooling3D : {...}
  VolumetricMaxPooling : {...}
  RNNReLU : {...}
  RNNTanh : {...}
  Tanh : {...}
  LogSoftMax : {...}
  BLSTM : {...}
  SpatialConvolution : {...}
  VolumetricBatchNormalization : {...}
  SpatialMaxPooling : {...}
  functional : 
    {
      bias2D_accGradParameters : function: 0x40663078
      Pooling_updateGradInput : function: 0x41bf3ec8
      Pooling_updateOutput : function: 0x41def828
      Convolution2D_accGradParameters : function: 0x4051fea0
      MaxPooling2D_updateGradInput : function: 0x40574408
      Convolution2D_updateOutput : function: 0x404cdad0
      bias2D_updateOutput : function: 0x419e7ea8
      AveragePooling2D_updateGradInput : function: 0x40574428
      AveragePooling2D_updateOutput : function: 0x405743e8
      MaxPooling2D_updateOutput : function: 0x405743c8
      Convolution2D_updateGradInput : function: 0x404b7180
    }
  GRU : {...}
  SpatialFullConvolution : {...}
  LSTM : {...}
  RNN : {...}
  BatchNormalization : {...}
  errcheck : function: 0x4056abf0
  VolumetricAveragePooling : {...}
  Sigmoid : {...}
  handle : cdata<struct cudnnContext *[?]>: 0x41cdcb40
  SpatialSoftMax : {...}
  _Pointwise : {...}
}
```

**Comment by slayerduck** [comment]
Sorry for my late response. The results for me are:

th> require 'cudnn'
{
  SpatialCrossEntropyCriterion : {...}
  VolumetricConvolution : {...}
  SpatialAveragePooling : {...}
  C : userdata: 0x400f9770
  benchmark : false
  ClippedReLU : {...}
  _Pooling : {...}
  SpatialBatchNormalization : {...}
  fastest : false
  toDescriptor : function: 0x400fa2e8
  ReLU : {...}
  SpatialCrossMapLRN : {...}
  convert : function: 0x405578d0
  version : 5103
  SoftMax : {...}
  TemporalConvolution : {...}
  getHandle : function: 0x400f9788
  SpatialLogSoftMax : {...}
  getSharedWorkspace : function: 0x400fa3a0
  _Pooling3D : {...}
  VolumetricMaxPooling : {...}
  RNNReLU : {...}
  RNNTanh : {...}
  Tanh : {...}
  LogSoftMax : {...}
  BLSTM : {...}
  SpatialConvolution : {...}
  VolumetricBatchNormalization : {...}
  SpatialMaxPooling : {...}
  functional :
    {
      bias2D_accGradParameters : function: 0x4055af58
      Pooling_updateGradInput : function: 0x40555960
      Pooling_updateOutput : function: 0x405558e8
      Convolution2D_accGradParameters : function: 0x4055a4a0
      MaxPooling2D_updateGradInput : function: 0x40555998
      Convolution2D_updateOutput : function: 0x4055af88
      bias2D_updateOutput : function: 0x4055ae70
      AveragePooling2D_updateGradInput : function: 0x405559b8
      AveragePooling2D_updateOutput : function: 0x40555940
      MaxPooling2D_updateOutput : function: 0x40555920
      Convolution2D_updateGradInput : function: 0x4055af20
    }
  GRU : {...}
  SpatialFullConvolution : {...}
  LSTM : {...}
  RNN : {...}
  BatchNormalization : {...}
  errcheck : function: 0x400fa2c0
  VolumetricAveragePooling : {...}
  Sigmoid : {...}
  handle : cdata<struct cudnnContext *[?]>: 0x41068200
  SpatialSoftMax : {...}
  _Pointwise : {...}
}

**Comment by nagadomi** [comment]
It seems that cudnn is correctly loaded.
Honestly, I have no idea why is `cudnn` not defined in `w2nn.load_model`. But I've updated `w2nn.lua`. Please retry with it.
If this issue cannot be solved, could you try with cudnn 5.0?  I am using cudnn 5.0.

**Comment by nagadomi** [comment]
Are you using LuaJIT 2.1?
```
$ luajit -v
LuaJIT 2.1.0-beta1 -- Copyright (C) 2005-2015 Mike Pall. http://luajit.org/
```
Lua5.1/5.2 has some issue, I think. #133

**Comment by slayerduck** [comment]
I was using Cudnn 5.0 before i tried upgrading to 5.1 to fix this error
Updated the w2nn.lua didn't produce different results. 

i'm using LuaJIT 2.0.4

**Comment by nagadomi** [comment]
Would you use LuaJIT 2.1?
If your torch env was builded with old [ezinstall](https://github.com/torch/ezinstall),
```
curl -s https://raw.githubusercontent.com/torch/ezinstall/master/clean-old.sh | bash
```
if you are manually installing luajit 2.0,
```
cd /path/to/luajit2.0_dir/
sudo make uninstall
```
And install torch with http://torch.ch/docs/getting-started.html
If you are already using this,
```
cd ~/torch
./clean.sh
./update.sh
./install-deps
./install.sh
```

**Comment by slayerduck** [comment]
I just removed the entire torch and reinstalled the whole bunch from scratch. Not sure if that was needed but it was easier for me to just try from 0.

Now version: LuaJIT 2.1.0-beta1

I then tried to convert and image with the new version and it errored out on the nvidia driver, so i then installed the latest version of the driver (367.35) and didn't touch the cudnn lib and then suddenly it worked! 

However the results are a bit dissapointing. With a test image -force_cudnn 1 is slower. Is that normal?

[waifu@waifu2x waifu2x-master]$ th waifu2x.lua -m noise -noise_level 1 -i 1.jpg -o output_image.png
output_image.png: 0.39446687698364 sec
[waifu@waifu2x waifu2x-master]$ th waifu2x.lua -m noise -noise_level 1 -i 1.jpg -o output_image.png -force_cudnn 1
output_image.png: 0.44212508201599 sec

**Comment by nagadomi** [comment]
Maybe first conversion is slow.
Try batch conversion or use `tools/benchmark.lua`.
```
find ./image_dir -name "*.png" > filelist.txt
mkdir out
time waifu2x -m scale -l filelist.txt -o out/%s.png -force_cudnn <0/1>
```
```
th tools/benchmark.lua -method scale -dir ./image_dir -force_cudnn <0/1>
```
`model1_time` is processing time that excludes image loading time.
